backend/eventloop/application_event_loop/application_event_runner_deployments.go

```diff
@@ -420,6 +432,54 @@ func (a applicationEventLoopRunner_Action) handleDeleteGitOpsDeplEvent(ctx conte
 
 }
 
+func removeFinalizerIfExist(ctx context.Context, k8sClient client.Client, gitopsDepl *managedgitopsv1alpha1.GitOpsDeployment, finalizer string) error {
+	if gitopsDepl == nil {
+		return nil
+	}
+
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		err := k8sClient.Get(ctx, client.ObjectKeyFromObject(gitopsDepl), gitopsDepl)
+		if err != nil {
+			if apierr.IsNotFound(err) {
+				return nil
+			}
+			return err
+		}
+
+		finalizers := removeItemFromSlice(finalizer, gitopsDepl.Finalizers)
+		if len(finalizers) != len(gitopsDepl.Finalizers) {
+			gitopsDepl.Finalizers = finalizers
+			return k8sClient.Update(ctx, gitopsDepl)
+		}
+
+		return nil
+	})
+}
+
+// we consider the GitOpsDeployment as deleted if:
+// 1. the GitOpsDeployment object is not found.
+// 2. the GitOpsDeployment is under deletion and the deletiontimestamp is set.
+func isGitOpsDeploymentDeleted(gitopsDepl *managedgitopsv1alpha1.GitOpsDeployment) bool {
+	return gitopsDepl == nil || gitopsDepl.DeletionTimestamp != nil
+}
+
+func removeItemFromSlice(item string, items []string) []string {
```

Comment:
Good catch! I forgot that slices are actually pointers to the underlying array and copying a slice and modifying it will update the original. To keep it simple I've updated the function to:
```
func removeItemFromSlice(item string, items []string) []string {
	result := []string{}
	for _, i := range items {
		if i != item {
			result = append(result, i)
		}
	}

	return result
}
```